opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1229,7 +1234,144 @@ namespace Opm {
+                    if (group.hasWell(well_name)) {
+                        auto& ws = well_state.well(well_name);
+                        rate = -tcalc.calcModeRateFromRates(ws.surface_rates);
+                        well_group_rate += rate;
+                        thp = ws.thp;
+                        min_thp = std::min(min_thp, thp);
+                        max_thp = std::max(max_thp, thp);
+                    }
+                }
 
+                const auto upbranch = network.uptree_branch(nodeName);
+                const auto it = node_pressures_.find((*upbranch).uptree_node());
+                const double nodal_pressure = it->second;
+                double well_group_thp = nodal_pressure;
+
+                if ((reportStepIdx <= 1) || (well_group_rate > 0.9*orig_target)) {
+                    std::array<double, 2> range;
+                    if (well_group_thp_prev_ == 0) {
+                        range = {0.9*min_thp, 1.1*max_thp};
+                    }
+                    else {
+                        range =  {0.9* this->well_group_thp_prev_, 1.1*this->well_group_thp_prev_}; 
+                    }
+
+                    double low, high;
+
+                    std::optional<double> approximate_solution;
+                    const double limit = 1.0E-3;
+                    local_deferredLogger.debug("Using brute force search to bracket the common THP");
+                    const bool finding_bracket = WellBhpThpCalculator::bruteForceBracketCommonTHP(mismatch, range, low, high, approximate_solution, limit, local_deferredLogger);
+
+
+                    if (approximate_solution.has_value()) {
+                        well_group_thp = *approximate_solution;
+                        double check = mismatch(well_group_thp);
+                        local_deferredLogger.debug("Approximate value found: mismatch = " +  std::to_string(check) + ", well_group_thp = " + std::to_string(well_group_thp));
+                    } else if (finding_bracket) {
+                        const double tolerance = 1.0E-3;
+                        const int max_iteration_solve = 100;
+                        int iteration = 0;
+                        well_group_thp = RegulaFalsiBisection<ThrowOnError>::
+                                         solve(mismatch, low, high, max_iteration_solve, tolerance, iteration);
+                        local_deferredLogger.debug(" bracket = [" + std::to_string(low) + ", " + std::to_string(high) + "], " +
+                                                   "iteration = " + std::to_string(iteration));
+                        double check = mismatch(well_group_thp);
+                        local_deferredLogger.debug(" mismatch = " +  std::to_string(check) + ", well_group_thp = " + std::to_string(well_group_thp));
+                    } else {
+                        local_deferredLogger.warning("Common THP solve failed due to bracketing failure");
```

Comment:
You can say here that the upstream nodal pressure is used for this nodal pressure for completeness.